### ISSUE TITLE
TypeScript compile error in typings for v3.1.2 

### ISSUE THREAD

**Issue by josundt** [aurelia/i18n#324]
**I'm submitting a bug report**

* **Library Version:**
3.1.2

* **Operating System:**
Windows [10]

* **Node Version:**
LTS

* **NPM Version:**
LTS

* **Language:**
TypeScript 4.0.2

**In **aurelia-i18n@3.1.2**, the _typings_ (aurelia-i18n/aurelia-i18n.d.ts) causes a TypeScript compile error.**

The error is related to the import statement for _i18next_ on **line 6** in _aurelia-i18n/aurelia-i18n.d.ts_:
```typescript
import { i18next } from 'i18next';
```  

This line was correct in the previous release (3.1.1) but has for some reason changed. It should still have been like before:
```typescript
import i18next from 'i18next';
```
_PS! I have **i18next@14.1.1** installed which comes with included typings, so I have _don't_ have the @types/i18next package installed as it should not be required._

When looking at the i18next typings, you see that i18next is a **default** export:
```typescript
declare namespace i18next {
    //[...]
}

declare const i18next: i18next.i18n;
export default i18next;
```

Not sure why this line changed since the 3.1.1 release, but it is a showstopper from upgrading in typescript projects.
Hopefully this is an easy fix that can be fixed quickly followed by an update released as soon as possible.

**Comment by zewa666** [comment]
Not really because it would force users to have additional settings in their tsconfig. Besides that while default exports are common in React and other frameworks they're not really seen as a good practice. We had this one popping up several times in existing issues.

**Comment by josundt** [comment]
**The current typings breaks the TypeScript compilation regardless of the tsconfig settings you use!**

I guess the tsconfig settings referred are `allowSyntheticDefaultImports` and/or `esModuleInterop` which allows using "default import syntax" similar to BabelJS for interoperability with CommonJS modules. I prefer not changing the defaults for these settings (false), I rather prefer the `import * as name from "module"` syntax for interoperability.

But the value of these tsconfig settings should not matter for i18next import, because i18next is exported as an **actual default export** - both in the ES module compilation and the CommonJS compilation - and should therefore be imported using default import syntax. (i18next >= 14.1.1).

_From i18next/dist/**es**/index.js (referenced in package.json "module" - the one used when bundling with webpack), line 14:_
```javascript
export default i18next;
```

_From i18next/dist/**commonjs**/index.js (referenced in package.json "main"), line 14:_
```javascript
exports.default = _i18next2.default;
```

When the i18next library is using _actual_ default exports, the import statement in the typings should also use default import syntax
```typescript
import i18next from i18next;
```
...and no special tsconfig settings should be required. 

**The current import statement in your typings**
```typescript
import { i18next } from i18next;
```
**will cause TypeScript compilation errors with any tsconfig settings.**

I have now tested in my project:
* I have fixed the import statement in node_modules/aurelia-i18n/dist/aurelia-i18n.d.ts to  `import 18next from i18next;`
* Then in my tsconfig.json, have then tried all combinations of `allowSyntheticDefaultImports` and `esModuleInterop`.

Compilation works regardless of the tsconfig settings (TypeScript 4.0.2).
But without this change in the typings file, compilation fails with any combination of the mentioned tsconfig settings.

**Comment by josundt** [comment]
PS! There seems to be a difference in the typings in the `@types/i18next` package used in your repository (v8.4.5) and the ones that come with `i18next` (v14.1.1) package that could be clarifying:

_From the `@types/i18next@8.4.5` typings:_  
```typescript
declare namespace i18next {
    // [...]
}

declare const i18next: i18next.i18n;
export = i18next;
```

_From `i18next@14.1.1` typings:_
```typescript
declare namespace i18next {
    // [...]
}

declare const i18next: i18next.i18n;
export default i18next;
```

The `@types/i18next` package has been **deprecated**, you should now use the one included in i18next instead. These typings also express correctly how i18next is actually exported - as a default export.

**Comment by zewa666** [comment]
Thx for looking into it. I will take a look asap and I'm sure we'll be rolling a fix out soon. Meanwhile if this is a blocker for you I'd recommend patch-package which allows you to safely navigate around the issue by patchfixing the d.ts file after npm installs.

**Comment by josundt** [comment]
Np. Sorry for causing all this work lately :-)
And sorry for adding as much text as I have in this case.
The compressed information is:
* the deprecated `@types/i18next` typings may have been vulnerable to different tsconfig settings - but the included typings with latest `i18next` should not have this problem.
* until the import is changed to `import i18next from 'i18next' in your typings, there will be compile errors for TypeScript projects - at least when @types/i18next is not installed
* What should be done: remove @types/i18next from the repo devDependencies, potentially update i18next to lastest (14.1.1) and inform users to uninstall @types/i18next.

**Comment by josundt** [comment]
Please let me know if you my contribution for this fix.
I realized I was involved in this earlier as well (#237).

**Comment by zewa666** [comment]
Aahh I wasn't sure but had the idea that the two of us already had contact with that issue. Thanks again and no absolutely no disturbance, it's me who's in debt with the low time.

If you can create a PR with your change it would be super helpful for testing. I should be able to take a look at it over the weekend.

**Comment by josundt** [comment]
Added a PR now that should fix the problem (#325).
It would be nice to have this released ASAP since the 3.1.2 release will be problematic for Aurelia TypeScript based projects. 

> A small unrelated tip I wanted to mention that could save you some time - since it seems we are both working on Intl/Jest/full-icu testing. I found that running tests using the command `cross-env NODE_ICU_DATA=node_modules/full-icu jest` failed for me when running multiple tests in parallel (Windows) - icu was not loaded in all child processes. I therefore had to add a Jest [globalsetup](https://jestjs.io/docs/en/configuration#globalsetup-string) file instead and set `process.env.NODE_ICU_DATA="node_modules/full-icu"` there...

**Comment by zewa666** [comment]
thanks again for the quick help, I was literally turning up VSCode to start looking at the issue. What a service ;)
I realized over at the Aurelia v2 version of the plugin we're already on 17.x thats why we didn't have the issue there. So thanks for the great catch.

By time, and once v2 alpha bubbles up I'll take the time to update the v1 plugins dependency as well and ship it with the next major. But for now we should be definitely good. Thanks again and after pinging @EisenbergEffect 3.1.3 should be on it's way soon.

**Comment by josundt** [comment]
@zewa666 Just testet out the 3.1.3 release. Looks like the typings file (aurelia-i18n.d.ts) file has not been updated!
PSI did not add in changes in the "dist" folder in my PR as I was instructed not to do that earlier - some CI process would take care of this folder. But in aurelia-i18n@3.1.3, it seems `npm run build` has not been run before releasing the package so it seems the dist folder is identical to previous release 🙁

**Comment by zewa666** [comment]
there was something totally strange since the change you did for 3.1.3 already landed previously and got released with 3.1.2 if you look at the repos commit history. I'm unsure how this happened but frankly when I just created a fresh new Webpack TS project and added aurelia-i18n I have no issues and all works. The i18next dependency is ^14.1.1 and no types/i18next present. So it seems ok to me, do you experience any issues?

**Comment by zewa666** [comment]
ok now I think I see the issue. If I try to inject I18N and then `instance.i18next.use` I get a type info. Now the built in typings export still via `export default i18next;` which is why we still need our script we had before that changes from named to default import.

God, the typings of i18next are really so messed up and these default exports are really giving me a grind. Can you verify that @josundt ?

**Comment by josundt** [comment]
I now realize that I have made a mistake.
The `fixI18nDefaultImport` function was actually fixing the problem, not causing it 😫 
The function was replacing `import { i18next } from 'i18next'` with `import i18next from 'i18next'` (desired), so it was a mistake from my side to remove it from the build script in my last PR 

But when aurelia-i18n@3.1.2 was built, for some reason the `fixI18nDefaultImport` function can not have been called, because the i18next import statement in the typings file was not fixed accordingly in the package release. 
When I updated the package to that version in my Aurelia project and then discovered the typings issue, I had a look at the build script assuming that the function was called during the package build.

Since I regretfully did not read the code in the function carefully, I assumed that `fixI18nDefaultImport` was doing **the exact opposite** of what it did, I thought it was replacing `import i18next from 'i18next'` with `import { i18next } from 'i18next'`, not the other way around. I obviously should have looked closer at the function, but when assuming that the function had run and that the typings file was a result of that, it was easy to jump to that conclusion after skimming the function's content.

Bottom line of this issue: The problem with the 3.1.2 release was that the `fixI18nDefaultImport` function for some reason never fixed the typings like expected. The question is why. This has not been a problem in preceding releases. 
AND: The function should NOT have been removed - which I unfortunately did previous PR before the 3.1.3 release. Really sorry about that... 

@zewa666 I think I have shared all the insights I have in this matter. Do you think you could figure this out from here and hopefully release yet another version - hopefully with the correct typings once and for all?

I am not familiar with all aspects of this package's build process, and I never used the `dts-bundle-generator` myself. In the packages I have made, I have normally generated typings simply by having an extra tsconfig file that inherits from the base one, but with 
```json
{ 
 "declaration": true,
 "declarationDir": "./dist/typings",
 "emitDeclarationOnly": true
}
```
Then, in 'package.json' I set the 'typings' property to e.g. 'dist/typings/aurelia-i18n.d.ts'.

It looks like the purpose of the `dts-bundle-generator` is to bundle all types into one d.ts file, but also to include only types used in the dependency graph of the exported types'; like some kind of "tree-shaking for types". I understand the rationale.

All `i18next` import statements in your typescript source code (e.g. './src/i18n.ts', './src/t/t-value-converter') use correct _default_ imports (`import i18next from 'i18next'). It is therefore strange to me that in the output of the `dts-bundle-generator` they have changed to (`import { i18next } from 'i18next'`.

That makes me question the reliability of this tool. I don't see any great benefits of bundling typings into one file. Maybe the "types tree-shaking" feature delivers a tiny performance benefit for the typescript compiler/language service, but maybe it's not worth it if like I suspect - modifies import statements etc.?

**Comment by zewa666** [comment]
Agree the `dts-bundle-generator` is an old relict from early Aurelia times. It certainly is going to be got rid of but was postponed to be done with the next major update.

So bottom line, the generated aurelia dts file should have `import i18next from 'i18next';` as it used to be with the fixup task (like in 3.1.1) so the actual change is merely to have an updated i18next dependency and the drop of the @types package right?

If so I'll create a fix for that and get it deployed.

When on to the update for the latest major I'll make sure these things get addressed in a better way. This is now being too often an issue.

**Comment by josundt** [comment]
@zewa666 That's exactly right 👍.  
Pleas also make sure that the typings fix is actually is applied to the typings in the dist folder before publishing the package, since it for some reason did not happen in the 3.1.2 release...

**Comment by doktordirk** [comment]
might be to add a typescript test. took me a little to follow the trail after getting the error after an `yarn upgrade`

**Comment by josundt** [comment]
@zewa666 Will you be able to fix this and make yet another release?
It should be more or less just to revert to the commit for the 3.1.2 release, but to figure out why the typings fix was not applied by the build preceding the package publishing of 3.1.2.

**Comment by zewa666** [comment]
hey guys, sorry that it took me so long. I've created the PR could you give it another look @josundt just to be safe :)

I guess that should really be it. Aside from the fix I've also made sure to update the jspm dependency which wasn't the same as the npm dependency.

**Comment by josundt** [comment]
@zewa666 I pulled your PR branch and looked at the code.
_After_ running `npm run build` the generated typings file is completely correct, but the file in the repo is incorrect _before_ running `npm run build` .
As mentioned I don't fully understand build/release process or why the build artifacts of the dist folder is not ignored in the repository, but if `npm run build` is executed before `npm publish` for the upcoming version; the package should end up completely correct 👍

**Comment by zewa666** [comment]
Released

**Comment by josundt** [comment]
It's been a bumpy road but it seems we finally we nailed it 😀
Thanks a lot!

**Comment by doktordirk** [comment]
@josundt the dist folder is not in gitignore since it contains the transpiled code for eg npm instalations. it shouldn't be commited (outside releases) since it's superficial and would make reviews 50times harder.
